velox/type/DecimalUtil.h

```diff
@@ -91,28 +116,126 @@ class DecimalUtil {
+      int32_t targetScale,
+      raw_vector<int32_t>& rows,
+      const velox::common::Filter& filter) {
+    int32_t numRows = rows.size();
+    bool isDense = (rows.back() - rows[0] == numRows - 1);
+    int32_t rowsCnt = 0;
+    if constexpr (std::is_same_v<T, int64_t>) { // Short Decimal
+      int startIndex = 0;
+      // short decimal can go fast path
+      if (!nullsPtr && process::hasAvx2()) {
+        constexpr int kBatchSize = xsimd::batch<int32_t>::size;
+        int halfBatch = kBatchSize / 2;
+        int32_t vecSize = numRows - numRows % kBatchSize;
+        auto targetScales = xsimd::batch<int64_t>::broadcast(targetScale);
+        for (int32_t i = 0; i < vecSize; i += kBatchSize) {
+          // deal with first half
+          auto fstValues = xsimd::load_aligned(values + i);
+          auto scalesVec = xsimd::load_aligned(scales + i);
+          int32_t matchMask = simd::toBitMask(scalesVec == targetScales);
+          if (matchMask != simd::allSetBitMask<int64_t>()) {
+            auto powersVec = simd::gather<int64_t, int64_t>(
+                kPowersOfTenShort, xsimd::abs(scalesVec - targetScales));
+            fstValues = xsimd::select(
+                scalesVec < targetScales,
+                fstValues * powersVec,
+                fstValues / powersVec);
+          }
+          // deal with second half
+          int32_t startRow = i + halfBatch;
+          auto secValues = xsimd::load_aligned(values + startRow);
+          scalesVec = xsimd::load_aligned(scales + startRow);
+          matchMask = simd::toBitMask(scalesVec == targetScales);
+          if (matchMask != simd::allSetBitMask<int64_t>()) {
+            auto powersVec = simd::gather<int64_t, int64_t>(
+                kPowersOfTenShort, xsimd::abs(scalesVec - targetScales));
+            secValues = xsimd::select(
+                scalesVec < targetScales,
+                secValues * powersVec,
+                secValues / powersVec);
+          }
+          int32_t fstMask = simd::toBitMask(filter.testValues(fstValues));
+          int32_t secMask = simd::toBitMask(filter.testValues(secValues));
+          int32_t mask = fstMask | secMask << halfBatch;
+
+          // save values and outputRows
+          if (mask) {
+            if (mask == simd::allSetBitMask<int32_t>()) {
+              xsimd::load_aligned(&rows[i]).store_aligned(&rows[rowsCnt]);
+              fstValues.store_aligned(decimals + rowsCnt);
+              rowsCnt += halfBatch;
+              secValues.store_aligned(decimals + rowsCnt);
+              rowsCnt += halfBatch;
+            } else {
+              if (isDense) {
+                auto fstRow = rows[rowsCnt];
+                (xsimd::load_unaligned(simd::byteSetBits(mask)) + fstRow)
+                    .store_aligned(&rows[rowsCnt]);
+              } else {
+                simd::filter(xsimd::load_aligned(&rows[i]), mask)
+                    .store_aligned(&rows[rowsCnt]);
+              }
+              if (fstMask) {
```

Comment:
Extract helper function for storing values.  Then you can break the data dependency a bit:
```
storeValues(fstMask, decimals + rowsCnt);
storeValues(secMask, decimals + rowsCnt + __builtin_popcount(fstMask));
```